Add and calculate coordinate container for off-rank neighbor nodes.

RyanWollaeger opened the pull request:
### Background

* The off-rank Dual_Ghost_Layout container can now see node indices local to the off-rank meshes that share a face with the MPI rank boundary nodes (effectively one node-layer into the off-rank mesh).
* However, to expedite computation over node-based stencils, the coordinate values of the nodes can be obtained with a similar communication pattern used to get the node indices on the off-rank mesh.

### Purpose of Pull Request

* Similar to how the off-rank node indices were obtained in PR #1078, add a allgatherv communication for the coordinates of the off-rank nodes sharing a face with the rank-boundary nodes.
  * This amounts to 4 doubles for each "corner element" (a cell adjacent to the current node and the other nodes of the cell's two faces adjacent to the current node). 
* [Fixes Redmine Issue #2457](https://rtt.lanl.gov/redmine/issues/2457)

### Description of changes

+ Create node-coordinate map that is 1-to-1 with map to node indexes in dual ghost layout.
+ Add specializations on double of allgatherv and determinate_allgatherv.
+ Use double determinate_allgatherv to get ghost neighbor node coordinates.
+ Use consistent loop variable names in loops serializing data for allgatherv.
+ Add tests of ghost coordinate values to tstDraco_Mesh_DD unit test.
+ Add mesh decomposition ASCII art schematics to tstDraco_Mesh_DD tests.

Note: this performs an additional allgatherv of four doubles between ranks
(two doubles for each coordinate adjacent to the node and its neighbor cell).

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis/Appveyor CI checks pass
  * [x] Code coverage does not decrease
  * [ ] [Valgrind test passes](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [ ] [Toss3 checks pass](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [ ] [Trinitite checks pass](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation

RyanWollaeger committed:
Add and calculate coordinate container for off-rank neighbor nodes.

+ Create node-coordinate map that is 1-to-1 with map to node indexes in dual ghost layout.
+ Add specializations on double of allgatherv and determinate_allgatherv.
+ Use double determinate_allgatherv to get ghost neighbor node coordinates.
+ Use consistent loop variable names in loops serializing data for allgatherv.
+ Add tests of ghost coordinate values to tstDraco_Mesh_DD unit test.
+ Add mesh decomposition ASCII art schematics to tstDraco_Mesh_DD tests.

Note: this performs an additional allgatherv of four doubles between ranks
(two doubles for each coordinate adjacent to the node and its neighbor cell).

RyanWollaeger commented:
Forgot about the `double == double` error; will replace these with `soft_equiv`.

clevelam commented:
@brryan This is another good PR to track. Ryan is ensuring we have access to the full dual mesh which will be required for DDMC on unstructured mesh.

RyanWollaeger committed:
Use soft_equiv for checking equality of doubles.

RyanWollaeger committed:
Merge branch 'develop' of github.com:lanl/Draco into dual_ghost_coord

KineticTheory commented:
The Appveyor CI failure isn't related to this PR.  Are you able to restart it?  If not, I'll build/test this PR in my environment to confirm that it is clean.  Let me know...

KineticTheory commented on src/c4/gatherv_pt.cc:
```diff
@@ -51,6 +51,8 @@ template void indeterminate_allgatherv<unsigned>(std::vector<unsigned> &outgoing
 template void determinate_allgatherv<unsigned>(std::vector<unsigned> &outgoing_data,
                                                std::vector<std::vector<unsigned>> &incoming_data);
+template void determinate_allgatherv<double>(std::vector<double> &outgoing_data,
+                                             std::vector<std::vector<double>> &incoming_data);
```
Is there a test in `c4` for this instantiation?  If not, can you add a simple test?

RyanWollaeger replied:
That is a good idea - will add a simple test for this instantiation.

RyanWollaeger replied:
Some unit tests for these `[in]determinate_allgatherv` instantiations have been added to `tstGatherScatter.cc`

RyanWollaeger replied:
Woops - need a `soft_equiv` for the test with `double`.

RyanWollaeger replied:
This is fixed now.

KineticTheory commented on src/mesh/Draco_Mesh.cc:
```diff
@@ -605,8 +632,15 @@ void Draco_Mesh::compute_node_to_cell_linkage(
+                                              coord_nbrs_per_serial_per_rank[rank][4 * i + 1]};
+      const std::array<double, 2> crd_nbr2 = {coord_nbrs_per_serial_per_rank[rank][4 * i + 2],
+                                              coord_nbrs_per_serial_per_rank[rank][4 * i + 3]};
+      ghost_coord_nbrs_per_rank[rank][global_node].push_back(std::make_pair(crd_nbr1, crd_nbr2));
```
Does `emplace_back` work here?

RyanWollaeger replied:
Yes - it should work here - thank you catching this!

RyanWollaeger replied:
These `push_back` calls have been replaced with `emplace_back`.

KineticTheory commented on src/mesh/test/tstDraco_Mesh_DD.cc:
```diff
@@ -20,6 +21,18 @@ using rtt_mesh_test::Test_Mesh_Interface;
+  //    |    |    |
+  //  0 ------------> x
+  //    0    1    2
+  //----------------------------------------------------------------------------------------------//
```
Nice!

RyanWollaeger replied:
Thanks!  These tests have a lot of hard-coded checks on the layouts, which can be difficult to read through.
Until the layout checks are more streamlined, a basic schematic of the decomposition seemed useful.

KineticTheory reviewed (approved):
LGTM.  Just a couple of comments.

RyanWollaeger commented:
@KineticTheory I can't seem to restart it from GitHub (I might be missing something however).
Tried creating an Appveyor account, but this did not seem to permit me to restart it (could not see how...).

KineticTheory commented:
> 
> 
> @KineticTheory I can't seem to restart it from GitHub (I might be missing something however).
> Tried creating an Appveyor account, but this did not seem to permit me to restart it (could not see how...).

I was able to restart it.  I forgot that I needed to log into the Appveyor system.  Let's see what it reports.

It still fails with `not an archive` error.  But I think I know the cause.  I'll need to make some changes to re-activate Appveyor CI.  For this PR, I'll test locally.

KineticTheory commented:
@RyanWollaeger BTW - the MSVC builds/tests are clean.

I'll put in a PR to fix the broken Appveyor CI.

RyanWollaeger committed:
Incorporate recommended improvements from KT.

+ Replace push_back with emplace_back in dual ghost map generation.
+ Add unit tests of [in]determinate_allgatherv specializations.
+ Fix else branch (missing) in logic for other gatherv unit tests.

RyanWollaeger committed:
Convert type to auto in unit test, to address clang-tidy result.

KineticTheory commented:
> > @KineticTheory I can't seem to restart it from GitHub (I might be missing something however).
> > Tried creating an Appveyor account, but this did not seem to permit me to restart it (could not see how...).
> 
> I was able to restart it. I forgot that I needed to log into the Appveyor system. Let's see what it reports.
> 
> It still fails with `not an archive` error. But I think I know the cause. I'll need to make some changes to re-activate Appveyor CI. For this PR, I'll test locally.

I was able to fix the Appveyor issue.  It seems to be running now.
